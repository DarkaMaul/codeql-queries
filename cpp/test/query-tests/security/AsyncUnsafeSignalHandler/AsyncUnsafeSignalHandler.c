#include "../../../include/libc/string_stubs.h"
#include "../../../include/libc/signal.h"
#include "../../../include/libc/stdlib.h"

void transitive_call() {
    printf("UNSAFE");
}
void transitive_call3() {
    int *x = (int*)malloc(16);
    free(x);
}
void transitive_call2() {
    transitive_call3();
}

// Signal handler function
void safe_handler(int sig) {
    if (sig == SIGALRM) {
        kill(1, SIGALRM);
    }
}

void unsafe_handler(int sig) {
    transitive_call();
}

void unsafe_handler2(int signo, siginfo_t *info, void *context) {
    if (signo == 3) {
        _Exit(0);
    }
    transitive_call2();
}

void unsafe_handler3(int signal) {
    transitive_call3();
}

int main() {
    // Register the safe signal handler
    if (signal(SIGALRM, safe_handler) == SIG_ERR) {
        perror("Unable to catch SIGALRM");
        _Exit(EXIT_FAILURE);
    }

    // Unsafe example
    if (signal(SIGALRM, unsafe_handler) == SIG_ERR) {
        perror("Unable to catch SIGALRM");
        _Exit(EXIT_FAILURE);
    }

    // Another unsafe example
    struct sigaction act = { 0 };
    act.sa_flags = SA_SIGINFO;
    act.sa_sigaction = &unsafe_handler2;
    if (sigaction(SIGSEGV, &act, NULL) == -1) {
        perror("sigaction");
        _Exit(EXIT_FAILURE);
    }

    struct sigaction act2 = { 0 };
    act2.sa_flags = SA_SIGINFO;
    act2.sa_handler = &unsafe_handler3;
    if (sigaction(SIGALRM, &act2, NULL) == -1) {
        perror("sigaction 2");
        _Exit(EXIT_FAILURE);
    }

    return 0;
}