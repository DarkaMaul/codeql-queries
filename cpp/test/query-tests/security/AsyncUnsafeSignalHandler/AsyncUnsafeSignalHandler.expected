| AsyncUnsafeSignalHandler.c:17:6:17:17 | safe_handler | is a non-trivial signal handler that may be using not async-safe functions. Registered by $@ | AsyncUnsafeSignalHandler.c:40:9:40:14 | call to signal | call to signal |
| AsyncUnsafeSignalHandler.c:23:6:23:19 | unsafe_handler | is a non-trivial signal handler that may be using not async-safe functions. Registered by $@ | AsyncUnsafeSignalHandler.c:46:9:46:14 | call to signal | call to signal |
